File: solutions/python3/0152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = nums[0]
        prevMin = nums[0]
        prevMax = nums[0]

        for i in range(1, len(nums)):
            min_ = prevMin * nums[i]
            max_ = prevMax * nums[i]
            prevMin = min(nums[i], min_, max_)
            prevMax = max(nums[i], min_, max_)
            ans = max(ans, prevMax)

        return ans

File: solutions/cpp/0152.cpp
class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    int ans = nums[0];
    int prevMin = nums[0];
    int prevMax = nums[0];

    for (int i = 1; i < nums.size(); ++i) {
      int min = prevMin * nums[i];
      int max = prevMax * nums[i];
      prevMin = std::min(nums[i], std::min(min, max));
      prevMax = std::max(nums[i], std::max(min, max));
      ans = std::max(ans, prevMax);
    }

    return ans;
  }
};